nengo/builder/optimizer.py

```diff
@@ -469,561 +759,137 @@ def _perform_merges_for_view_subset(
+                if _OpMerger.is_mergeable(op2, tomerge):
+                    tomerge.add(op2)
 
-        return op_replacements, sig_replacements
+            if len(tomerge.ops) > 1:
+                self.merge(tomerge.ops)
 
-    def _merge(self, merge, op_replacements, sig_replacements):
-        """Merges the operators in `merge` and updates `op_replacements` and
-        `sig_replacements` with the required replacements of operators and
-        signals.
+    def merge(self, ops):
+        """Merges the given operators.
 
-        It will also update the internal list of merged operators to prevent
-        further merges on the same operators before all required operators and
-        signals have been replaced.
+        This method will also update ``op_replacements``, ``sig_replacements``,
+        and the internal list of merged operators to prevent further merges
+        on the same operators before all required operators and signals have
+        been replaced.
         """
-        merged_op, merged_sig = self.merge_ops(merge)
-        self._merged.update(merge)
-        for op in merge:
-            op_replacements[op] = merged_op
+        merged_op, merged_sig = _OpMerger.merge(ops)
+        self.merged.update(ops)
+        for op in ops:
+            self.op_replacements[op] = merged_op
             # Mark all operators referencing the same signals as merged
             # (even though they are not) to prevent them from getting
             # merged before their signals have been updated.
             for s in op.all_signals:
-                self._merged.update(self._sig2op[s])
-        sig_replacements.update(merged_sig)
-
-    def _update_dg(self, op_replacements):
-        """Updates the dependency graph.
-
-        This involves doing all operator replacements in `op_replacements` and
-        ensuring that dependencies are preserved.
-        """
-        dg = {}
-        for op in op_replacements:
-            if op_replacements[op] not in dg:
-                dg[op_replacements[op]] = set()
-            for x in self.dg[op]:
-                dg[op_replacements[op]].add(op_replacements[x])
-        self.dg.clear()
-        self.dg.update(dg)
-
-    @staticmethod
-    def _map_onto_op_signals(fn, ops):
-        """Maps function `fn` on all signals of all operators in the sequence
-        `ops`.
-        """
-        for op in ops:
-            for key in dir(op):
-                sig = getattr(op, key)
+                self.merged.update(self.sig2op[s])
+        self.sig_replacements.update(merged_sig)
+
+    def finalize(self):
+        """Finalizes merges done during the pass."""
+
+        for old, new in iteritems(self.op_replacements):
+            assert old is not new
+            if new not in self.dg:
+                self.dg[new] = set()
+            for dep in self.dg[old]:
+                self.dg[new].add(self.op_replacements.get(dep, dep))
```

Comment:
I think these two lines can be replaced by
`self.dg[new].update(self.dg[old])` because the replacement is done in the next loop anyways.